compiler/rustc_error_messages/src/lib.rs

```diff
@@ -357,6 +357,17 @@ impl<S: Into<String>> From<S> for DiagnosticMessage {
     }
 }
 
+/// A workaround for "good path" ICEs when formatting types in disables lints.
+///
+/// Delays formatting until `.into(): DiagnosticMessage` is used.
+pub struct DelayDm<F>(pub F);
```

Comment:
Will this go away after the migration to translatable messages?

Reply:
Yes, this is the intention.

Reply:
Should we add a rustc lint to detect when we're using `Ty: Display` without wrapping it in this? (Maybe not to address in this PR, but I know I'm bound to make that mistake myself at some point in the future.)

Reply:
I think we should add a lint, yes.